crates/bevy_ecs/src/archetype.rs

```diff
@@ -244,17 +245,17 @@ impl Edges {
         archetype_id: ArchetypeId,
         bundle_status: impl Into<Box<[ComponentStatus]>>,
         required_components: impl Into<Box<[RequiredComponentConstructor]>>,
-        added: impl Into<Box<[ComponentId]>>,
-        existing: impl Into<Box<[ComponentId]>>,
+        added: Vec<ComponentId>,
+        existing: impl IntoIterator<Item = ComponentId>,
     ) {
         self.insert_bundle.insert(
             bundle_id,
             ArchetypeAfterBundleInsert {
                 archetype_id,
                 bundle_status: bundle_status.into(),
                 required_components: required_components.into(),
-                added: added.into(),
-                existing: existing.into(),
+                added_len: added.len(),
+                inserted: added.into_iter().chain(existing).collect(),
```

Comment:
I'm not sure if Vec specializes this, but using Extend or append and then converting into the boxed slice is both cleaner and may occasionally elide the forced allocation here.

Reply:
> I'm not sure if Vec specializes this, but using Extend or append and then converting into the boxed slice is both cleaner and may occasionally elide the forced allocation here.

I went back and forth on this.  The trouble is that a naïve approach using `extend` or `append` uses `reserve` instead of `reserve_exact` and might over-allocate, causing two allocations instead of just one.  And then I went down the rabbit hole of trying to figure out if we could also re-use the allocation from `existing` if the one from `added` wasn't the right size, before deciding I was making it too complicated.  

Let me try the version with an explicit `reserve_exact`.